network/wsNetwork.go

```diff
@@ -1264,6 +1295,123 @@ func (wn *WebsocketNetwork) meshThread() {
+	need := desired - numOutgoingTotal
+	if need <= 0 {
+		return false
+	}
+	// get more than we need so that we can ignore duplicates
+	newAddrs := wn.phonebook.GetAddresses(desired + numOutgoingTotal)
+	for _, na := range newAddrs {
+		if na == wn.config.PublicAddress {
+			// filter out self-public address, so we won't try to connect to outselves.
+			continue
+		}
+		gossipAddr, ok := wn.tryConnectReserveAddr(na)
+		if ok {
+			wn.wg.Add(1)
+			go wn.tryConnect(na, gossipAddr)
+			need--
+			if need == 0 {
+				break
+			}
+		}
+	}
+	return true
+}
+
+// checkExistingConnectionsNeedDisconnecting check to see if existing connection need to be dropped due to
+// performance issues and/or network being stalled.
+func (wn *WebsocketNetwork) checkExistingConnectionsNeedDisconnecting() bool {
+	// we already connected ( or connecting.. ) to  GossipFanout peers.
+	// get the actual peers.
+	outgoingPeers := wn.outgoingPeers()
+	if len(outgoingPeers) < wn.config.GossipFanout {
+		// reset the performance monitor.
+		wn.connPerfMonitor.Reset([]Peer{})
+		return wn.checkNetworkAdvanceDisconnect()
+	}
+
+	if !wn.connPerfMonitor.ComparePeers(outgoingPeers) {
+		// different set of peers. restart monitoring.
+		wn.connPerfMonitor.Reset(outgoingPeers)
+	}
+
+	// same set of peers.
+	peerStat := wn.connPerfMonitor.GetPeersStatistics()
+	if peerStat == nil {
+		// performance metrics are not yet ready.
+		return wn.checkNetworkAdvanceDisconnect()
+	}
+
+	// update peers with the performance metrics we've gathered.
+	var leastPerformingPeer *wsPeer = nil
+	for _, stat := range peerStat.peerStatistics {
+		wsPeer := stat.peer.(*wsPeer)
+		wsPeer.peerMessageDelay = stat.peerDelay
+		wn.log.Infof("network performance monitor - peer '%s' delay %d first message portion %d%%", wsPeer.GetAddress(), stat.peerDelay, int(stat.peerFirstMessage*100))
+		if wsPeer.throttledOutgoingConnection && leastPerformingPeer == nil {
+			leastPerformingPeer = wsPeer
+		}
+	}
+	if leastPerformingPeer == nil {
+		return wn.checkNetworkAdvanceDisconnect()
+	}
+	wn.disconnect(leastPerformingPeer, disconnectLeastPerformingPeer)
+	wn.connPerfMonitor.Reset([]Peer{})
+
+	return true
+}
+
+// checkNetworkAdvanceDisconnect is using the lastNetworkAdvance indicator to see if the network is currently "stuck".
+// if it's seems to be "stuck", a randomally picked peer would be disconnected.
+func (wn *WebsocketNetwork) checkNetworkAdvanceDisconnect() bool {
+	lastNetworkAdvance := wn.getLastNetworkAdvance()
+	if time.Now().UTC().Sub(lastNetworkAdvance) < cliqueResolveInterval {
+		return false
+	}
+	outgoingPeers := wn.outgoingPeers()
+	if len(outgoingPeers) == 0 {
+		return false
+	}
+	if wn.numOutgoingPending() > 0 {
+		// we're currently trying to extend the list of outgoing connections. no need to
+		// disconnect any existing connection to free up room for another connection.
+		return false
+	}
+	var peer *wsPeer
+	disconnectPeerIdx := crypto.RandUint63() % uint64(len(outgoingPeers))
```

Comment:
GossipFanout = 4 by default. In case of sporadic disconnect of some other node we have quite high probability of waiting `2 x cliqueResolveInterval` on partitioned network

Reply:
Relays are expected to be relatively reliable, so there shouldn't be too many sporadic disconnection. The waiting 2 x cliqueResolveInterval is the fallback for agreement protocol not making any progress. Otherwise, the number of outgoing connections would be examined once a minute.